src/extensions/default/SVGCodeHints/main.js

```diff
@@ -0,0 +1,307 @@
+        if (isSVGEnabled && editor.getModeForSelection() === "image/svg+xml") {
+            this.editor = editor;
+            this.tagInfo = SVGUtils.getTagInfo(this.editor, this.editor.getCursorPos());
+            
+            if (this.tagInfo && this.tagInfo.tokenType) {
+                return true;
+            }
+        }
+        return false;
+    };
+    
+    /**
+     * Returns a list of hints that are available in the current context,
+     * or null if there are no hints available. 
+     *
+     * @param {string} implicitChar A character that the user typed in the hinting session.
+     * @return {hints: Array.<string>, match: string, selectInitial: boolean, handleWideResults: boolean}
+     */
+    SVGCodeHints.prototype.getHints = function (implicitChar) {
+        var hints = [], query, tagInfo, i, attrs = [], tagAttr, tagSpecificAttrs, isMultiple;
+        
+        tagInfo  = SVGUtils.getTagInfo(this.editor, this.editor.getCursorPos());
+        this.tagInfo = tagInfo;
+        
+        if (tagInfo && tagInfo.tokenType) {
+            query = tagInfo.token.string.substr(0, tagInfo.offset).trim();
+        
+            if (tagInfo.tokenType === SVGUtils.TokenTag) {
+                for (i in tagData.tags) {
+                    if (tagData.tags.hasOwnProperty(i) && i.indexOf(query) === 0) {
+                        hints.push(i);
+                    }
+                }
+            } else if (tagInfo.tokenType === SVGUtils.TokenAttr) {
+                if (!tagData.tags[tagInfo.tagName]) {
+                    return null;
+                }
+                // Get attributes.
+                attrs = getTagAttributes(tagInfo.tagName);
+                hints = attrs.filter(function (value) {
+                    if (value.indexOf(query) === 0 && tagInfo.exclusionList.indexOf(value) === -1) {
+                        return true;
+                    }
+                });
+            } else if (tagInfo.tokenType === SVGUtils.TokenValue) {
+                tagAttr = tagInfo.tagName + "/" + tagInfo.attrName;
+                tagSpecificAttrs = attributeData[tagAttr];
+                
+                if (!tagData.tags[tagInfo.tagName] && !(attributeData[tagInfo.attrName] || tagSpecificAttrs)) {
+                    return null;
+                }
+                
+                // Get attribute options.
+                if (tagSpecificAttrs) {
+                    // Prefer tag/attribute for specific tags.
+                    attrs = tagSpecificAttrs.attribOptions || attributeData[tagSpecificAttrs.alias].attribOptions;
+                    isMultiple = tagSpecificAttrs.multiple;
+                } else if (attributeData[tagInfo.attrName]) {
+                    // Else Use general options for attributes.
+                    attrs = attributeData[tagInfo.attrName].attribOptions || attributeData[attributeData[tagInfo.attrName].alias].attribOptions;
```

Comment:
Would be better if you'd store `attributeData[tagInfo.attrName]` in a variable to simplify this line.